### PR TITLE
Added std::string_view<>

### PR DESCRIPTION
**Pull request by TurkeyMan**

### CODE REVIEW

**Comment by TurkeyMan** [comment]
```
string_view.cpp
test\stdcpp\src\string_view.cpp(1) : fatal error C1083: Cannot open include file: 'string_view': No such file or directory
```
Really? We're rolling ancient compilers here? :(

**Comment by rainers** [comment]
> Really? We're rolling ancient compilers here? :(

The auto-tester builds with VS2013, string_view was added in VS2017.

**Comment by TurkeyMan** [comment]
Removed the Win64 test. I'll add it back whenever the auto-tester can run it.
I'm consuming it locally for now, so it will continue to work.

**Comment by TurkeyMan** [comment]
This is super-useful, because can finally pass D-strings between D/C++ with no local hacks!

**Comment by TurkeyMan** [comment]
> cc: error: unrecognized command line option ‘-std=c++17’

Oh FFS... these testers are all vintage machines! >_<

We just have to trust that this one works, since CI can't run any of the tests (they work for me locally!)

File: src/core/stdcpp/string_view.d [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    {
+        ///
+        this(const(T)[] str) nothrow @trusted @nogc                 { _Mydata = str.ptr; _Mysize = str.length; }
+
+        ///
+        size_type size() const nothrow @safe @nogc                  { return _Mysize; }
+        ///
+        const(T)* data() const nothrow @safe @nogc                  { return _Mydata; }
+        ///
+        const(T)[] as_array() const inout @trusted @nogc            { return _Mydata[0 .. _Mysize]; }
```
[the remark this answers is not in the document]
Comment (TurkeyMan) [reply src/core/stdcpp/string_view.d]: It's at the top.

[the remark this answers is not in the document]
Comment (TurkeyMan) [reply src/core/stdcpp/string_view.d]: All the containers do this.

**Comment by TurkeyMan** [comment]
Difficult to test, because very few CI machines support C++17, including the Windows CI machines...

**Comment by thewilsonator** [comment]
Hmm, clearly some of them do. Why do we not have Appveyor for druntime? Then you could disable the autotester's windows builds and rely on that instead.

**Comment by TurkeyMan** [comment]
I think we might just need to merge this on faith until all the CI machines get with the now.. I use it locally, people will report bugs if it doesn't work, but honestly, it will likely get VERY low usage in the short term, since it's not that common even in C++ yet.

**Comment by thewilsonator** [comment]
Ahh I see that you haven't enabled the test.

**Comment by TurkeyMan** [comment]
This is pretty critical for D interaction though; it allows you to pass string slices to/from D. This is the C++ slice ABI that people keep complaining we need to have.

**Comment by thewilsonator** [comment]
Heh, I'd love to get 
https://github.com/dlang/dmd/pull/8120 merged.

**Comment by TurkeyMan** [comment]
Yeah, it would be nice... but if we have `std::string_view<>` and `std::span<>`, do we really need that hack?
I guess it's still nice for C++ to link to D code, rather than the case of D linking to C++ code (where these are useful).

**Comment by TurkeyMan** [comment]
Remove `auto_merge` for a minute, I just want to give it a good once-over, and run the tests comprehensively since I rebased.

**Comment by thewilsonator** [comment]
Done. Ping me when you're ready.

**Comment by jacob-carlborg** [comment]
> Heh, I'd love to get dlang/dmd#8120 merged.

The problem is that the ABI for D strings and a struct with a corresponding layout doesn't match on Windows. D strings are passed in registers but not structs. So we would need to change the ABI as well in addition to the mangling.

**Comment by TurkeyMan** [comment]
No. They will perform the same as if you remove D's slice ABI optimisation. I'm saying, don't change the slice ABI, just use another struct instead (perhaps these).

**Comment by jacob-carlborg** [comment]
I was thinking to change it only for `extern(C++)` functions.

**Comment by TurkeyMan** [comment]
And extern(C)?

**Comment by jacob-carlborg** [comment]
> And extern(C)?

I didn't think of that. There's no mangling for C functions so the mangling problem doesn't exist there. But the ABI mismatch does, I guess.

**Comment by TurkeyMan** [comment]
VS LTO is not supported by any 3rd party toolchains.

**Comment by thewilsonator** [comment]
Of course,*grumble,grumble*. It does work fine for LDC/Clang and I presume GDC/G++.

**Pull request merged**